ambry-commons/src/main/java/com/github/ambry/commons/FutureUtils.java

```diff
@@ -0,0 +1,95 @@
+ */
+
+package com.github.ambry.commons;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import static com.github.ambry.utils.Utils.*;
+
+
+/**
+ * A collection of utilities that expand the usage of {@link CompletableFuture}. Inspired from
+ * https://jar-download.com/artifacts/org.apache.flink/flink-runtime_2.11/1.7.0/source-code/org/apache/flink/runtime/concurrent/FutureUtils.java
+ */
+public class FutureUtils {
+
+  /**
+   * Retry the given operation the given number of times in case of a failure.
+   *
+   * @param <T> type of the result
+   * @param operation to executed
+   * @param retries if the operation failed
+   * @param retryPredicate Predicate to test whether an exception is retryable
+   * @param retryErrorConsumer
+   * @param finalErrorConsumer
+   * @return Future containing either the result of the operation or a {@link Exception} if error occurred.
+   */
+  public static <T> CompletableFuture<T> retry(final Supplier<CompletableFuture<T>> operation, final int retries,
+      Predicate<Throwable> retryPredicate, Consumer<Throwable> retryErrorConsumer,
+      Consumer<Throwable> finalErrorConsumer) {
+    final CompletableFuture<T> resultFuture = new CompletableFuture<>();
+    retryOperation(resultFuture, operation, retries, retryPredicate, retryErrorConsumer);
+    resultFuture.whenComplete((t, throwable) -> {
+      if (throwable != null) {
+        finalErrorConsumer.accept(throwable);
+      }
+    });
+    return resultFuture;
+  }
+
+  /**
+   * Helper method which retries the provided operation in case of a failure.
+   * @param <T> type of the future's result
+   * @param resultFuture to complete
+   * @param operation to retry
+   * @param retries until giving up
+   * @param retryPredicate Predicate to test whether an exception is retryable
+   * @param retryErrorConsumer
+   */
+  private static <T> void retryOperation(final CompletableFuture<T> resultFuture,
+      final Supplier<CompletableFuture<T>> operation, final int retries, Predicate<Throwable> retryPredicate,
+      Consumer<Throwable> retryErrorConsumer) {
+
+    if (!resultFuture.isDone()) {
+      final CompletableFuture<T> operationFuture = operation.get();
+
+      operationFuture.whenCompleteAsync((t, throwable) -> {
+        if (throwable != null) {
+          if (throwable instanceof CancellationException) {
+            resultFuture.completeExceptionally(throwable);
+          } else {
+            throwable = extractFutureExceptionCause(throwable);
+            if (!retryPredicate.test(throwable)) {
+              resultFuture.completeExceptionally(throwable);
+            } else if (retries > 0) {
+              retryErrorConsumer.accept(throwable);
+              retryOperation(resultFuture, operation, retries - 1, retryPredicate, retryErrorConsumer);
```

Comment:
It seems like this doesn't do any backoff. Is that okay for this use case?

Reply:
I was wondering if it should be okay since this is per our current behavior. This is for use case in `ADAuthBasedStorageClient` where we are refreshing our token and retrying blob operations immediately when we receive authorization failure due to token expiry. 

May be I can expand the utility method to take in backoff duration and retry accordingly when we use it for other use cases.

Reply:
This makes sense. ADAuthBasedStorageClient may be retired soon anyways.